### DISCUSSION TITLE
solved:Find a way to automatically change [$global:foo] to [$script:foo] in ps1 script?

### DISCUSSION THREAD

**Discussion by kasini3000** [PowerShell/PowerShell#16756] [Q&A] [answered]
I used powershell to build a better container management tool than kubernetes,project named 【k4t】
My ideal is: user management containers do not need to use YAML, everyone needs to write powershell scripts.
k4t is very important, k4t can make powershell great again and again.

The k4t engine executes many user callback ps1 scripts. 

Is there a way to prevent users from using global variables, or
automatically limit global variables to script scope? 
eg: 
$global:foo  ---> $script:foo

Restricts the current thread, runspace child thread, can only read the global scope.

I have thought about run the script with subprocess, 
but each powershell process takes up too much memory, at least 120mb. This is not a good way.
can you reduce the memory usage of the powershell process?

**Comment by iSazonov** [comment, marked as answer]
Did you consider runspaces?

**Reply by kasini3000** [reply]
Yes, if only runspace is used ,current thread is not used, the global scope will not be polluted.
You inspired me, thank you.
